### PR TITLE
1159 add diffusive abm and smm

### LINKED ISSUE #1159: Add diffusive ABM and SMM
### Feature description

Add the diffusive abm including the quadwell potential and the smm from https://github.com/reneSchm/memilio/tree/example-hybrid-omp to main.

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### PR DESCRIPTION
**Pull request by jubicker**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- add diffusive ABM (model, simulation, parameters and quadwell potential)
- add SMM (model, simulation, parameters)
- add tests for both models
- add examples for both models

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [ ] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

closes #1159

### CODE REVIEW

File: cpp/models/d_abm/quad_well.h [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        , m_non_moving_states(non_moving_states)
+        , m_number_transitions(static_cast<size_t>(Status::Count), Eigen::MatrixXd::Zero(4, 4))
+    {
+        for (auto& agent : populations) {
+            mio::unused(agent);
+            assert(is_in_domain(agent.position));
+        }
+        for (auto& r : rates) {
+            m_adoption_rates.emplace(std::forward_as_tuple(r.region, r.from, r.to), r);
+            //m_adoption_rates[{r.region, r.from, r.to}] = r;
```
Comment (reneSchm) [inline cpp/models/d_abm/quad_well.h:66]:
```suggestion
```

File: cpp/models/d_abm/model.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    using Implementation::move;
+    using Implementation::time_point;
+
+    using Status = typename Implementation::Status;
+    using Agent  = typename Implementation::Agent;
+
+    inline constexpr void check_constraints() const
+    {
+    }
+};
```
Comment (reneSchm) [inline cpp/models/d_abm/model.h:47]: Whoever would write all of this without a singular comment?
```suggestion

/**
 * @brief Wrap an implementation of a diffusive ABM so it can be run by the d_abm::Simulation.
 * Uses the CRTP. See comments on using statements for expected function signatures.
 * @tparam Implementation A class implementing all functions and types marked with the using keyword in Model.
 */
template <class Implementation>
class Model : public Implementation
{
public:
    /// Use the constructors defined by the Implementation
    using Implementation::Implementation;

    /**
     * @brief Set the status of an agent.
     * Expected signature: `void adopt(Agent&, const Status&)`
     */
    using Implementation::adopt;

    /**
     * @brief Calculate the current adoption rate of an agent from its status to the given one.
     * Expected signature: `double adoption_rate(const Agent&, const Status&)`
     */
    using Implementation::adoption_rate;

    /**
     * @brief Change the Position of an Agent, depending on its state, the current time and step size.
     * Expected signature: `void move(const double, const double, Agent&)`
     * The first argument is time, the second step size.
     */
    using Implementation::move;

    /**
     * @brief Get the Implementations RNG.
     * Expected signature: `mio::RandomNumberGenerator& get_rng()`
     */
    using Implementation::get_rng;

    /**
     * @brief Aggregate the population by their Status for the simulation result.
     * Expected signature: `Eigen::VectorXd time_point()`
     */
    using Implementation::time_point;

    /// @brief The status of an agent.
    using Status = typename Implementation::Status;
    /// @brief An agent is expected to contain at least a status and a position.
    using Agent = typename Implementation::Agent;

    /// @brief Empty function for compatability with MEmilio.
    inline constexpr void check_constraints() const
    {
    }
};```

**Review by reneSchm — CHANGES_REQUESTED** [review]
I seem to barely document code I write under a deadline. I will continue this with more suggestions later, but feel free to edit them - or add your own documentation, there is a lot to do...

The directory structure and naming looks pretty good. I have not checked the tests yet

File: cpp/tests/test_d_abm_model.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    R,
+    D,
+    Count
+
+};
+
+TEST(TestQuadWell, well)
+{
+    auto well = well_index(Eigen::Vector2d{-1, 1});
+    EXPECT_EQ(well, size_t(0));
```
Comment (reneSchm) [inline cpp/tests/test_d_abm_model.cpp:46]: Test 1-3 as well?

File: cpp/tests/test_d_abm_model.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+{
+    QuadWellModel<InfectionState>::Agent a1{Eigen::Vector2d{-1, 1}, InfectionState::S};
+    QuadWellModel<InfectionState>::Agent a2{Eigen::Vector2d{-1.2, 1}, InfectionState::I};
+    QuadWellModel<InfectionState> qw({a1, a2}, {{InfectionState::S,
+                                                 InfectionState::E,
+                                                 mio::dabm::Region(0),
+                                                 0.1,
+                                                 {InfectionState::C, InfectionState::I},
+                                                 {1, 0.5}}});
+    EXPECT_EQ(qw.adoption_rate(a1, InfectionState::E), 0.025);
```
Comment (reneSchm) [inline cpp/tests/test_d_abm_model.cpp:68]: Other branches that are not checked are
- the rate for a2 to E is 0
- another isolated infected does not influence the rate
- the rate of a1 scales correctly with more nearby infected and non-infected

File: cpp/tests/test_d_abm_model.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+{
+    QuadWellModel<InfectionState>::Agent a1{Eigen::Vector2d{-1.2, 1}, InfectionState::S};
+    QuadWellModel<InfectionState>::Agent a2{Eigen::Vector2d{-1.2, 1}, InfectionState::I};
+    QuadWellModel<InfectionState> qw({a1, a2}, {}, 0.1, 0., {InfectionState::I});
+    qw.move(0, 0.1, a1);
+    qw.move(0, 0.1, a2);
+    EXPECT_EQ(a1.position[0], -0.9888);
+    EXPECT_EQ(a1.position[1], 1);
+    EXPECT_EQ(a2.position[0], -1.2);
+    EXPECT_EQ(a2.position[1], 1);
```
Comment (reneSchm) [inline cpp/tests/test_d_abm_model.cpp:81]: This should not work well without mocking, as move uses a normal distributed value.

File: cpp/models/d_abm/quad_well.h [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            std::find(m_non_moving_regions.begin(), m_non_moving_regions.end(), old_well) ==
+                m_non_moving_regions.end()) {
+            Position p = {mio::DistributionAdapter<std::normal_distribution<double>>::get_instance()(m_rng, 0.0, 1.0),
+                          mio::DistributionAdapter<std::normal_distribution<double>>::get_instance()(m_rng, 0.0, 1.0)};
+
+            agent.position      = agent.position - dt * grad_U(agent.position) + (m_sigma * std::sqrt(dt)) * p;
+            const auto new_well = well_index(agent.position);
+            if (old_well != new_well) {
+                m_number_transitions[static_cast<size_t>(agent.status)](old_well, new_well)++;
+            }
```
Comment (reneSchm) [inline cpp/models/d_abm/quad_well.h:126]: Do we want to keep these counters?

File: cpp/models/smm/parameters.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+template <class Status>
+struct AdoptionRates {
+    using Type = std::vector<AdoptionRate<Status>>;
+    const static std::string name()
+    {
+        return "AdoptionRates";
+    }
+};
+
+template <class Status>
```
Comment (reneSchm) [inline cpp/models/smm/parameters.h:61]:
```suggestion
/**
 * @brief Struct defining a possible regional transition in a Model based on Poisson Processes.
 * @tparam Status An infection state enum.
 */
template <class Status>
```

File: cpp/tests/test_d_abm_model.cpp (hunk 5 above)
Comment (jubicker) [reply cpp/tests/test_d_abm_model.cpp:81]: It does because sigma is 0, therefore the movement is only defined by the function gradient.

Comment (reneSchm) [reply cpp/tests/test_d_abm_model.cpp:81]: Oh, I missed that. Can you add a comment that sigma=0? It is not that obvious from the constructor call.

File: cpp/models/d_abm/quad_well.h (hunk 6 above)
Comment (jubicker) [reply cpp/models/d_abm/quad_well.h:127]: I would say yes. They might be helpful for output analysis.

File: cpp/tests/test_smm_model.cpp [hunk 8]
```diff
@@ -156,16 +163,121 @@ TEST(TestSMMSimulation, advance)
...
+
+    //Simulate for 30 days
+    auto sim = mio::smm::Simulation(model, 0.0, 0.1);
+    sim.advance(30.);
+    //As model populations are all zero only t0 and tmax should be logged
+    EXPECT_EQ(sim.get_result().get_num_time_points(), 2);
     EXPECT_EQ(sim.get_result().get_last_time(), 30.);
 }
+
+TEST(TestSMMSimulation, covergence)
```
Comment (reneSchm) [inline cpp/tests/test_smm_model.cpp:209]:
```suggestion
TEST(TestSMMSimulation, convergence)
```

File: cpp/tests/test_smm_model.cpp [hunk 9]
```diff
@@ -156,16 +163,121 @@ TEST(TestSMMSimulation, advance)
...
+    }
+    //The number of transitions from region 0 to region 1 should be approx. rate * pop in region 0
+    double rel_diff1 =
+        std::abs(rate * pop1 - std::abs(std::accumulate(transitions1.begin(), transitions1.end(), 0.0) / num_runs1)) /
+        (rate * pop1);
+    double rel_diff2 =
+        std::abs(rate * pop2 - std::abs(std::accumulate(transitions2.begin(), transitions2.end(), 0.0) / num_runs2)) /
+        (rate * pop2);
+
+    EXPECT_GE(rel_diff1, rel_diff2);
```
Comment (reneSchm) [inline cpp/tests/test_smm_model.cpp:282]: Why do you compare the two models with each other rather than with the expected result? The tolerance for comparing with the result is probably fairly large, but I think it would be better than comparing two "random" values.

Also, why the greater equal? If both models converge, reldiff1 < rel_diff2 should be possible, right? Maybe use one model, and compare the simulated rate against the expected rate, with some tolerance (with can be fairly large).

Comment (reneSchm) [reply cpp/tests/test_smm_model.cpp:282]: Moreover, should this be a random test perhaps? So that it checks not only that one specific seed works. It is okay for a random test to possibly fail, if it is very unlikely that it does. You can run it a couple 1000 times to check that the tolerance is large enough for most edge cases.

**Review by reneSchm — CHANGES_REQUESTED** [review]

File: cpp/tests/test_d_abm_model.cpp (hunk 4 above)
Comment (reneSchm) [reply cpp/tests/test_d_abm_model.cpp]: Can you add descriptions to the test what it is you check? Otherwise it is hard to tell what is going on here.

This is true for all added tests. For example, in TestQuadWell.well(_index ?) one comment at the top like "check that well_index returns indices in the correct order" would be enough.
Here, in TestQuadWell.adoptionRate, you could write down what the agents and rates are used for, and/or explain what the EXPECTs do. Like "check that a1 is affected by a2, but not the for off a3" for the first EXPECT call, and so on.

File: cpp/tests/test_smm_model.cpp (hunk 9 above)
Comment (jubicker) [reply cpp/tests/test_smm_model.cpp:282]: What I do here is check whether the simulated transition rate converges to the expected transition rate for num_runs, pop_size to infty. And I realize this by checking whether the relative difference between the simulated number of transitions and the expected number of transitions gets smaller if I increase the number of runs as well as the number of agents.

I can also only use one model and compare the simulated and expected number of transitions, but what would be an appropriate tolerance? 1%, 50%? An how many runs do we require to meet that tolerance?

File: cpp/memilio/epidemiology/adoption_rate.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ * the sum over all "influences" scaled by their respective "factors".
+ * @tparam Status An infection state enum.
+ */
+template <class Status>
+struct AdoptionRate {
+    Status from; // i
+    Status to; // j
+    mio::regions::Region region; // k
+    ScalarType factor; // gammahat_{ij}^k
+    std::vector<Influence<Status>> influences;
```
[the remark this answers is not in the document]
Comment (jubicker) [reply cpp/memilio/epidemiology/adoption_rate.h:54]: Now we do

File: cpp/models/d_abm/parameters.h [hunk 11]
```diff
@@ -21,41 +21,10 @@
...
-};
-
-template <class Status>
-struct AdoptionRates {
-    using Type = std::vector<AdoptionRate<Status>>;
-    const static std::string name()
-    {
-        return "AdoptionRates";
-    }
-};
```
[the remark this answers is not in the document]
Comment (jubicker) [reply cpp/models/d_abm/parameters.h:28]: We decided to keep this file and add a comment on why it is empty.

File: cpp/models/d_abm/quad_well.h [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+                m_non_moving_states.end() &&
+            std::find(m_non_moving_regions.begin(), m_non_moving_regions.end(), old_well) ==
+                m_non_moving_regions.end()) {
+            Position p = {mio::DistributionAdapter<std::normal_distribution<double>>::get_instance()(m_rng, 0.0, 1.0),
+                          mio::DistributionAdapter<std::normal_distribution<double>>::get_instance()(m_rng, 0.0, 1.0)};
+
+            agent.position      = agent.position - dt * grad_U(agent.position) + (m_sigma * std::sqrt(dt)) * p;
+            const auto new_well = well_index(agent.position);
+            if (old_well != new_well) {
+                m_number_transitions[static_cast<size_t>(agent.status)](old_well, new_well)++;
```
Comment (reneSchm) [inline cpp/models/d_abm/quad_well.h:159]: Can you add an agent (or change ones position) in the move test, so this line gets covered? Maybe check as well that number transitions = 0,...,0,1,0,...,0

File: cpp/tests/test_smm_model.cpp [hunk 13]
```diff
@@ -242,42 +259,14 @@ TEST(TestSMMSimulation, covergence)
...
-        (rate * pop1);
-    double rel_diff2 =
-        std::abs(rate * pop2 - std::abs(std::accumulate(transitions2.begin(), transitions2.end(), 0.0) / num_runs2)) /
-        (rate * pop2);
-
-    EXPECT_GE(rel_diff1, rel_diff2);
+    double mean_trans = std::accumulate(transitions.begin(), transitions.end(), 0.0) / num_runs;
+    double rel_dev    = std::abs(expected_num_trans - mean_trans) / expected_num_trans;
+    //The relative deviation should not exceed 15%
+    EXPECT_LE(rel_dev, 0.15);
```
Comment (reneSchm) [inline cpp/tests/test_smm_model.cpp:271]: Nice, 15% is honestly better than I expected.

**Review by reneSchm — APPROVED** [review]
